go-controller/pkg/clustermanager/healthcheck/controller.go

```diff
@@ -0,0 +1,426 @@
+	if err != nil && !util.IsAnnotationNotSetError(err) {
+		return fmt.Errorf("failed to parse node %s subnets annotation %v", node.Name, err)
+	}
+
+	mgmtIPs := make([]net.IP, len(nodeSubnets))
+	for i, subnet := range nodeSubnets {
+		mgmtIPs[i] = util.GetNodeManagementIfAddr(subnet).IP
+	}
+
+	update := true
+	nodeInfo, probed := c.isProbedWithNodeInfo(nodeName)
+	if probed {
+		less := func(a, b net.IP) bool { return a.String() < b.String() }
+		update = !cmp.Equal(nodeInfo.ips, mgmtIPs, cmpopts.SortSlices(less))
+	}
+
+	if !update {
+		// already probing and no config update, bail out
+		return nil
+	}
+
+	if len(mgmtIPs) == 0 {
+		// no proper config, stop probing and assume unreachable out of
+		// precaution
+		updated := c.cancelProbeWithState(nodeName, UNREACHABLE)
+		if updated {
+			klog.Infof("Stopping health state probing for node %s: node management IPs unknown", nodeName)
+			c.informConsumers(nodeName)
+		}
+		return nil
+	}
+
+	nodeInfo.ips = mgmtIPs
+
+	// add/update probing task
+	klog.Infof("Starting or updating health state probing for node %s on IPs %s", nodeName, util.StringSlice(mgmtIPs))
+	c.startProbeWithNodeInfo(nodeName, nodeInfo)
+
+	return nil
+}
+
+func (c *controller) informConsumers(node string) {
+	c.consumerLock.RLock()
+	defer c.consumerLock.RUnlock()
+	// note we don't query the interest of the consumers as that might be racey,
+	// it is up to the consumers to appropriately filter out uninteresting
+	// events
+	for _, consumer := range c.consumers {
+		consumer.HealthStateChanged(node)
+	}
+}
+
+func (c *controller) hasInterestedConsumers(node string, names *[]string) bool {
+	c.consumerLock.RLock()
+	defer c.consumerLock.RUnlock()
+	for _, consumer := range c.consumers {
+		if names != nil {
+			*names = append(*names, consumer.Name())
+		}
+		if consumer.MonitorHealthState(node) {
+			return true
+		}
+	}
+	return false
+}
+
+func (c *controller) isProbedWithNodeInfo(node string) (nodeInfo, bool) {
+	c.RLock()
+	defer c.RUnlock()
+	nodeInfo, exists := c.nodeInfo[node]
+	return nodeInfo, exists
+}
+
+func (c *controller) startProbeWithNodeInfo(node string, nodeInfo nodeInfo) {
+	c.Lock()
+	defer c.Unlock()
+
+	// cancel any previous task and schedule a new one in case config
+	// changed
+	if nodeInfo.cancelTask != nil {
+		nodeInfo.cancelTask()
+	}
+
+	ctx, cancel := context.WithCancel(c.ctx)
+	nodeInfo.cancelTask = cancel
+	c.nodeInfo[node] = nodeInfo
+
+	task := &probeTask{
+		node:   node,
+		client: healthcheck.NewEgressIPHealthClient(node),
+		ctx:    ctx,
+	}
+
+	// schedule task with some jitter
+	j := time.Duration(rand.Intn(jitter)) * time.Millisecond
+	c.probeTaskQueue.Push(task, j)
+}
+
+func (c *controller) setNodeStateAndRetask(task *probeTask, new HealthState) bool {
+	on := time.Now()
+	c.Lock()
+	defer c.Unlock()
+
+	// check if task was cancelled after probing
+	if task.cancelled() {
+		task.client.Disconnect()
+		return false
+	}
+
+	updated := c.setNodeState(task.node, new, on, false)
```

Comment:
avoid using "new" as its also a builtin in golang. Maybe newHealthState or something